README.md
# Bloc REST API

<p>
<a href="https://pub.dev/packages/bloc_rest_api"><img src="https://img.shields.io/pub/v/bloc_rest_api" alt="Pub"></a>
<a href="https://github.com/muhammadidrees/bloc_rest_api/actions"><img src="https://github.com/muhammadidrees/bloc_rest_api/actions/workflows/publish.yml/badge.svg" alt="build"></a>
<a href="https://app.codecov.io/gh/muhammadidrees/bloc_rest_api"><img src="https://codecov.io/gh/muhammadidrees/bloc_rest_api/branch/master/graph/badge.svg" alt="codecov"></a>
<a href="https://github.com/dart-lang/lints#migrating-from-packagepedantic"><img src="https://img.shields.io/badge/style-pedantic-40c4ff.svg" alt="style: pedantic"></a>
<a href="https://opensource.org/licenses/MIT"><img src="https://img.shields.io/badge/license-MIT-purple.svg" alt="License: MIT"></a>

</p>

A flutter package to easily integrate and manage REST APIs. Now all you need to do is create a model class and all the rest is taken care of. :)

> Before using the package fimiliarize yourself with [bloc library](https://bloclibrary.dev/#/)

__FYI__: This package works on all platforms (Android, iOS, Web, Linux, Windows, Mac).

## Usage

To demonstrate the usage of the package we'll be using [JsonPlaceHolder](https://jsonplaceholder.typicode.com/)

First, we need to do add `bloc_rest_api` to the dependencies of the `pubspec.yaml`

```yaml
dependencies:
  bloc_rest_api: <latest version>
```

Next, we need to install it:

```sh
# Dart
pub get

# Flutter
flutter packages get
```

Now create a model class for the data that you want to fetch from the internet. This can easily be done by using online tools such as [QuickType](https://app.quicktype.io/)

> Note: Don't forget to use [Equatable](https://pub.dev/packages/equatable) or any other comparision method to compare the models as it is the decision maker when it comes to switching state in bloc

In our case we'll be creating a model for Post.

```dart
import 'package:equatable/equatable.dart';

class PostModel extends Equatable {
  PostModel({
    this.userId,
    this.id,
    this.title,
    this.body,
  });

  final int userId;
  final int id;
  final String title;
  final String body;

  factory PostModel.fromJson(Map<String, dynamic> json) => PostModel(
        userId: json['userId'] == null ? null : json['userId'],
        id: json['id'] == null ? null : json['id'],
        title: json['title'] == null ? null : json['title'],
        body: json['body'] == null ? null : json['body'],
      );

  Map<String, dynamic> toJson() => {
        'userId': userId == null ? null : userId,
        'id': id == null ? null : id,
        'title': title == null ? null : title,
        'body': body == null ? null : body,
      };

  @override
  List<Object> get props => [userId, id];
}
```

Now create a provider for RequestCubit of type model (in our case `PostModel`) and pass in the fromMap function i.e. the fucntion that converts the incoming json to model.

```dart
MultiBlocProvider(
  providers: [
    // for single model
    BlocProvider(
      create: (context) => RequestCubit<PostModel>(
        fromMap: (json) => PostModel.fromJson(json),
      ),
    ),
    // for list of posts simply update type and fromMap method
    BlocProvider(
      create: (context) => RequestCubit<List<PostModel>>(
        fromMap: (json) =>
            List<PostModel>.from(json.map((x) => PostModel.fromJson(x))),
      ),
    ),
  ],
  child: AppView(),
);
```

To use the `getRequest` or `postRequest` methods simply use the `context.read()` funtion like:

```dart
context.read<RequestCubit<PostModel>>().getRequest(
      baseUrl: 'https://jsonplaceholder.typicode.com/',
      handle: 'posts/1',
    );
```

You can also enable logs for requests in order to view logs happening for a specific function by setting `enableLogs` flag to `true`.

```dart
context.read<RequestCubit<PostModel>>().getRequest(
      ...
      enableLogs: true,
    );
```

Want to use local json strings? Don't worry the package got you covered.

```dart
context.read<RequestCubit<PostModel>>().localRequest(
      'json string here',
    );
```

> Note: You can always convert local json file to string using `loadString('path/to/asset')` from `import 'package:flutter/services.dart'`

In case you are dealing with complex data and want to handle the mapping yourself you can you the request method that takes a function that returns Future of type T as parameter.

```dart
Future<PostModel> fetchAlbum() async {
  final response =
      await http.get('https://jsonplaceholder.typicode.com/posts/1');

  if (response.statusCode == 200) {
    return PostModel.fromJson(jsonDecode(response.body));
  } else {
    // Any exception thrown will emit failure state
    throw Exception('Failed to load album');
  }
}

...
    context.read<RequestCubit<PostModel>>().request(fetchAlbum());
...
```

Finally react on the states by using either `BlocBuilder`, `BlocListner` or `BlocConsumer` method.

States can have one of 4 status distinguished as state.status:
     
1. RequestStatus.empty: Initial State
2. RequestStatus.loading: Loading State
3. RequestStatus.success: Success State
4. RequestStatus.error: Error State
 
__Note:__ The state retains the data on status change and is only changed either on the success state or by explicitly calling the emptyCubit method. This features allows you to show previous data even on API error.

Heres is an example builder method for PostModel:

```dart
BlocBuilder<RequestCubit<PostModel>, RequestState<PostModel>>(
  builder: (context, state) {
    switch (state.status) {
      case RequestStatus.empty:
        return Center(child: Text('Press the button to get some data'));

      case RequestStatus.loading:
        return Center(child: CircularProgressIndicator());

      case RequestStatus.success:
        return Center(
          child: Padding(
            padding: const EdgeInsets.all(32.0),
            child: Text(state.model.toString()),
          ),
        );

      case RequestStatus.failure:
        return Center(child: Text(state.errorMessage));
      default:
        return Container();
    }
  },
),
```

#### ApiConfig

`baseUrl`, `header` and `responseTimeOut` can also be configured globally for all `RequestCubit` instances via `ApiConfig`

```dart
ApiConfig.baseUrl = '...';
ApiConfig.header = {'...': ''};
ApiConfig.responseTimeOut = Duration(...);
```

#### Other Userful Methods

You can use the updateModel method to update the state model manually.

```dart
context.read<RequestCubit<PostModel>>().updateModel(model);
```

Also you can empty or reinitialize the state using empty() method.

```dart
context.read<RequestCubit<PostModel>>().emtpy();
```

### Presisting REST APIs

Sometimes you want to presist the api data. For that you can simply replace the `RequestCubit` to `HydratedRequestCubit`.

Just make sure to provide fromMap and toMap functions in the provider.
```dart
BlocProvider(
  create: (context) => HydratedRequestCubit<PostModel>(
    fromMap: (json) => PostModel.fromJson(json),
    toMap: (model) => model.toJson(),
  ),
),
```

And also initialize the storage
```dart
void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  HydratedBloc.storage = await HydratedStorage.build();
  runApp(App());
}
```

And now you're state will never go away. :)
## Maintainers

- [Muhammad Idrees](https://github.com/muhammadIdrees)

CHANGELOG.md
# 1.0.0-nullsafety.0

- **BREAKING**: migrated to null safety
- feat!: upgrade Dart SDK constraints to `>=2.12.0-0 <3.0.0`
- chore: updated flutter version in CI to latest

# 0.4.0+1

- chore: github action publish issue resolution

# 0.4.0

- feat: added functionality to read local json
- feat: added timeOut reference for request function to override global behaviour
- fix: assertions being caught
- fix: removed extra bracket from log
- docs: added example for local read, log feature and timeOut
- docs: added badges :)
- docs: updated old examples
- refactor: achieved 100% code coverage
- refactor: added padentic linter
- refactor: avoided dublication for hydrated cubit and extended the existing cubit instead
- chore: removed unnecessary code and imports

# 0.3.2

- feat: added option to enable log for a request function
- feat: used copyWith to allow cubit to retain model data on error and loading states
- feat: added format exception if unable to convert json to model
- chore: removed request model since it's no longer being used

# 0.3.1

- docs: spell corrections
- refactor: added pedantic as linter for package
# 0.3.0

- feat: use of copy with to retain model data on state change
- chore: bumped flutter_bloc to `flutter_bloc: ^6.1.3`
- chore: bumped meta to `^1.2.2`
- docs: added not for copywith change

# 0.2.0

- Added a request function to handle Future requests

# 0.1.1

- Unused import removed

# 0.1.0

- Initial Release

example/lib/main.dart
import 'dart:convert';

import 'package:bloc_rest_api/bloc_rest_api.dart';
import 'package:flutter/material.dart';
import 'package:flutter_bloc/flutter_bloc.dart';
import 'package:http/http.dart' as http;
import 'models/models.dart';

void main() {
  // Specify the base url
  ApiConfig.baseUrl = "https://jsonplaceholder.typicode.com/";
  // Specify base header for response
  ApiConfig.header = {"Content-Type": "application/json"};
  runApp(App());
}

class App extends StatelessWidget {
  App({Key key}) : super(key: key);

  // This widget is the root of the application.
  @override
  Widget build(BuildContext context) {
    return MultiBlocProvider(
      providers: [
        // for single model
        BlocProvider(
          create: (context) => RequestCubit<PostModel>(
            fromMap: (json) => PostModel.fromJson(json),
          ),
        ),
        BlocProvider(
          create: (context) => HydratedRequestCubit<PostModel>(
            fromMap: (json) => PostModel.fromJson(json),
            toMap: (model) => model.toJson(),
          ),
        ),
        // for list of posts simply update type and fromMap method
        BlocProvider(
          create: (context) => RequestCubit<List<PostModel>>(
            fromMap: (json) =>
                List<PostModel>.from(json.map((x) => PostModel.fromJson(x))),
          ),
        ),
      ],
      child: AppView(),
    );
  }
}

class AppView extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: HomePage(title: 'Example App'),
    );
  }
}

class HomePage extends StatelessWidget {
  HomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(title),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          // context.read<RequestCubit<PostModel>>().request(fetchAlbum());
          context.read<RequestCubit<PostModel>>().getRequest(
                handle: 'posts/1',
              );
        },
        child: Icon(Icons.add),
      ),
      body: BlocConsumer<RequestCubit<PostModel>, RequestState<PostModel>>(
        listener: (context, state) {
          if (state.status == RequestStatus.failure) {
            Scaffold.of(context).showSnackBar(
              SnackBar(content: Text(state.errorMessage)),
            );
          }
        },
        builder: (context, state) {
          switch (state.status) {
            case RequestStatus.empty:
              return Center(child: Text("Press the button to get some data"));

            case RequestStatus.loading:
              return Center(child: CircularProgressIndicator());

            case RequestStatus.success:
            case RequestStatus.failure:
              return Center(
                child: Padding(
                  padding: const EdgeInsets.all(32.0),
                  child: Text(state.model.toString()),
                ),
              );

            default:
              return Container();
          }
        },
      ),
    );
  }
}

Future<PostModel> fetchAlbum() async {
  final response =
      await http.get('https://jsonplaceholder.typicode.com/posts/1');

  if (response.statusCode == 200) {
    return PostModel.fromJson(jsonDecode(response.body));
  } else {
    // Any exception thrown will emit failure state
    throw Exception('Failed to load album');
  }
}
